### PR TITLE
[NO-1466] Specify a supplier to buy from in market contract

### PR DESCRIPTION
**Pull request by amiecorso**
Adds another sale endpoint to the market contract, `swapFromSpecificSupplier`, that allows specification of a supplier address from which to buy.  Some refactoring to get better code reuse. Between `swap` and `swapFromSpecificSupplier`, which now differ only in the ways that they determine the allocated supply.

### CODE REVIEW

File: contracts/FIFOMarket.sol [hunk 1]
```diff
@@ -267,6 +372,92 @@ contract FIFOMarket is
...
+
+  /**
+   * Determines the removal ids and amounts to fill the given purchase quantity, sourcing only
+   * from a single supplier.
+   */
+  function _allocateSupplySingleSupplier(
+    uint256 certificateAmount,
+    address supplier
+  )
+    internal
```
Comment (jaycenhorton) [inline contracts/FIFOMarket.sol:386]:
```suggestion
    private
```

File: contracts/FIFOMarket.sol [hunk 2]
```diff
@@ -267,6 +372,92 @@ contract FIFOMarket is
...
+   *
+   */
+  function _fulfillOrder(
+    uint256 certificateAmount,
+    address recipient,
+    uint256 numberOfRemovals,
+    uint256[] memory ids,
+    uint256[] memory amounts,
+    address[] memory suppliers
+  ) internal {
```
Comment (jaycenhorton) [inline contracts/FIFOMarket.sol:460]:
```suggestion
  ) private {
```

File: test/FIFOMarket.test.ts [hunk 3]
```diff
@@ -1186,4 +1186,143 @@ describe('FIFOMarket', () => {
...
+        supplierInitialNoriBalance.add(purchaseAmount)
+      );
+      expect(investor1FinalNoriBalance).to.equal(investor1InitialNoriBalance);
+      const sources = await certificate.sources(0);
+      expect(sources.length).to.equal(2);
+      const decodedRemovalIds = await Promise.all(
+        sources.map((source) => removal.unpackRemovalIdV0(source.removalId))
+      );
+      for (const decodedId of decodedRemovalIds)
+        expect(decodedId.supplierAddress).to.equal(hre.namedAccounts.supplier);
```
Comment (jaycenhorton) [inline test/FIFOMarket.test.ts:1249]:
```suggestion
      expect(decodedRemovalIds.map(e => e.supplierAddress)).to.deep.equal(
        Array.from({ length: decodedRemovalIds.length })
        .fill(hre.namedAccounts.supplier)
      );
```

**Review by jaycenhorton — APPROVED** [review]
Very clean :100:

File: contracts/FIFOMarket.sol [hunk 4]
```diff
@@ -267,6 +372,92 @@ contract FIFOMarket is
...
+        .queueByVintage[vintage]
+        .length();
+    }
+    require(totalNumberOfRemovalsForSupplier > 0, "Market: Not enough supply");
+    uint256 remainingAmountToFill = certificateAmount;
+    uint256[] memory ids = new uint256[](totalNumberOfRemovalsForSupplier);
+    uint256[] memory amounts = new uint256[](totalNumberOfRemovalsForSupplier);
+    uint256 numberOfRemovals = 0;
+    // TODO (Gas Optimization): Declare variables outside of loop
+    for (uint256 i = 0; i < totalNumberOfRemovalsForSupplier; i++) {
```
[the remark this answers is not in the document]
Comment (amiecorso) [reply contracts/FIFOMarket.sol:412]: I don't think so.... they iterate a different number of times and do different things inside the loop

**Review by jaycenhorton — APPROVED** [review]

**Pull request merged**